lenient/include/module.h

```diff
@@ -0,0 +1,77 @@
+/*
+ * A module for providing utilities for module management.
+ *
+ * To have visibility to `this` and `this_private` of any module during
+ * compilation, 'function context' function definitions are placed behind
+ * macros.
+ *
+ * TODO write what is expected from the file that generates functions using the
+ * macros defined here.
+ *
+ * TODO implement tests.
+ */
+
+#ifndef MODULE_H_DEFINED
+#define MODULE_H_DEFINED
+
+#include <stddef.h>
+#include <stdint.h>
+
+#define MODULE_GENERATE_CONTEXT_FUNCTIONS(struct_public, struct_full) \
+    MODULE_GENERATE_DECLARATIONS(struct_public)                       \
+    MODULE_GENERATE_LOAD_CONTEXT(struct_public, struct_full)          \
+    MODULE_GENERATE_UNLOAD_CONTEXT                                    \
+    MODULE_GENERATE_VALIDATE_CONTEXT
+
+#define MODULE_GENERATE_DECLARATIONS(struct_public)                \
+    static int32_t module_load_context(struct_public* new_this);   \
+    static void module_unload_context(void);                       \
+    static void module_validate_context(void);
```

Comment:
Please implement `module_exit_on_unloaded_context`. This could call `module_validate_context_loaded` and exit depending on the return value.

Reply:
Implemented in: 588f106b271b2b6927bfef585428636fdb35ea2a